src/fifelse.c

```diff
@@ -343,3 +343,194 @@ SEXP fcaseR(SEXP na, SEXP rho, SEXP args) {
   UNPROTECT(nprotect);
   return ans;
 }
+
+SEXP topnR(SEXP vec, SEXP n, SEXP dec) {
+	int nprotect = 0;
+	int64_t i, j, idx = 0;
+	int len0 = asInteger(n);
+	const int64_t len1 = xlength(vec);
+	
+	if (isS4(vec) && !INHERITS(vec, char_nanotime)) {
+		error("S4 class objects (excluding nanotime) are not supported.");
+	}
+	if (len0 > 1000) {
+		error("Function 'topn' is not built for large value of 'n'. The algorithm is made for small values. Please prefer the 'order' function if you want to proceed with such large value.");
+	}
+	if (len0 > len1) {
+		warning("'n' is larger than length of 'vec'. 'n' will be set to length of 'vec'.");
+		len0 = len1;
+	}
+	if (len0 < 1) {
+		error("Please enter a positive integer larger or equal to 1.");
+	}
+	if (!IS_TRUE_OR_FALSE(dec)) {
+		error("Argument 'decreasing' must be TRUE or FALSE and length 1.");
+	}
+	
+	const bool vdec = LOGICAL(dec)[0];
+	SEXPTYPE tvec = TYPEOF(vec);
+	SEXP ans = PROTECT(allocVector(INTSXP, len0)); nprotect++;
+	int *restrict pans = INTEGER(ans);
+	int tmp;
+
+	if (vdec) {
+		switch(tvec) {
+		case INTSXP: {
+			const int *restrict pvec = INTEGER(vec);
+			int min_value = pvec[0];
```

Comment:
IINM we haven't already returned length-0 input, so this will be out-of-memory access

Reply:
Not sure to understand. Do you mean we need to check that `vec` is of non null length? If yes, I think you are right I might need to add a check.

Reply:
Would you be able to provide an example that fails? Happy to correct it if needed.